### PR TITLE
[FLINK-7065] Separate the flink-streaming-java module from flink-clients

### PR DESCRIPTION
**Pull request by XuPingyong**

### CODE REVIEW

**Comment by XuPingyong** [comment]
@tillrohrmann , could you please spend some time on reviewing this PR?

**Comment by fhueske** [comment]
Hi @XuPingyong, @tillrohrmann is not available for the next two weeks.

File: flink-clients/src/main/java/org/apache/flink/client/RemoteExecutor.java [hunk 1]
```diff
@@ -106,8 +113,14 @@ public RemoteExecutor(InetSocketAddress inet, Configuration clientConfiguration,
 			List<URL> jarFiles, List<URL> globalClasspaths) {
 		this.clientConfiguration = clientConfiguration;
 		this.jarFiles = jarFiles;
+		for (URL jarFileUrl : jarFiles) {
+			try {
+				JobWithJars.checkJarFile(jarFileUrl);
+			} catch (IOException e) {
+				throw new RuntimeException("Problem with jar file " + jarFileUrl, e);
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/RemoteExecutor.java]: Can we make this a checked exception?

File: flink-clients/src/main/java/org/apache/flink/client/program/OptimizerPlanEnvironment.java [hunk 2]
```diff
@@ -37,8 +37,11 @@
 
 	private FlinkPlan optimizerPlan;
 
+	private StreamPlanEnvironment streamPlanEnvironment;
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/OptimizerPlanEnvironment.java]: can this be final?

File: flink-clients/src/main/java/org/apache/flink/client/program/PreviewPlanEnvironment.java [hunk 3]
```diff
@@ -38,6 +38,12 @@
 
 	Plan plan;
 
+	StreamPlanEnvironment streamPlanEnvironment;
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PreviewPlanEnvironment.java]: final?

File: flink-contrib/flink-connector-wikiedits/pom.xml [hunk 4]
```diff
@@ -37,7 +37,7 @@ under the License.
 	<dependencies>
 		<dependency>
 			<groupId>org.apache.flink</groupId>
-			<artifactId>flink-streaming-java_${scala.binary.version}</artifactId>
+			<artifactId>flink-clients_${scala.binary.version}</artifactId>
```
Comment (tillrohrmann) [inline flink-contrib/flink-connector-wikiedits/pom.xml]: Why does this module need a dependency on `flink-clients`? Is it being executed?

File: flink-core/src/main/java/org/apache/flink/api/common/Executor.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * A Executor execute a Flink program's job.
+ *
+ * <p>The specific implementation (such as the org.apache.flink.client.LocalExecutor
+ * and org.apache.flink.client.RemoteExecutor) determines where and how to run the dataflow.
+ * The concrete implementations of the executors are loaded dynamically, because they depend on
+ * the full set of all runtime classes.</p>
+ */
+@Internal
+public interface Executor {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/Executor.java]: Can we rename it? It clashes with `java.util.concurrent.Executor`. Maybe something like `ProgramExecutor` or so.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutorFactory.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	// ------------------------------------------------------------------------
+	//  Executor Factories
+	// ------------------------------------------------------------------------
+	
+	/**
+	 * Creates an executor that runs locally in a multi-threaded environment.
+	 * 
+	 * @return A local executor.
+	 */
+	public T createLocalExecutor(Configuration configuration) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutorFactory.java]: Can we make this a static method? Then we would not have to create a factory object.

File: flink-streaming-java/pom.xml [hunk 7]
```diff
@@ -52,7 +52,7 @@ under the License.
 
 		<dependency>
 			<groupId>org.apache.flink</groupId>
-			<artifactId>flink-clients_${scala.binary.version}</artifactId>
+			<artifactId>flink-optimizer_${scala.binary.version}</artifactId>
```
Comment (tillrohrmann) [inline flink-streaming-java/pom.xml]: can we remove the `flink-optimizer` dependency? Maybe we have to move `FlinkPlan` and `StreamingPlan` out of `flink-optimizer` for that.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamGraphExecutor.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	 * to explicitly start the executor for all programs where multiple dataflow parts
+	 * depend on each other. Otherwise, the previous parts will no longer
+	 * be available, because the executor immediately shut down after the execution.</p>
+	 *
+	 * @param streamGraph The streamGraph to execute.
+	 * @return The execution result, containing for example the net runtime of the program, and the accumulators.
+	 *
+	 * @throws Exception Thrown, if job submission caused an exception.
+	 */
+	JobExecutionResult executeStreamGraph(StreamGraph streamGraph) throws Exception;
```
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamGraphExecutor.java]: Could the thrown exception be of type `FlinkException`?

File: flink-contrib/flink-connector-wikiedits/pom.xml (hunk 4 above)
Comment (XuPingyong) [inline flink-contrib/flink-connector-wikiedits/pom.xml]: It is executed in WikipediaEditsSourceTest. In the latest commit, this module still depends on flink-streaming-java, and also flink-clients of test scope.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamGraphExecutor.java (hunk 8 above)
Comment (XuPingyong) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamGraphExecutor.java]: Like PlanExecutor， Exception is still be throwed, which may be mixed with multi exception types.

**Comment by XuPingyong** [comment]
Thanks @tillrohrmann , I have updated this PR according to your advice.

**Pull request closed without merging**